### PR TITLE
[4.3] Inject the database into workflow

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Ongoing effort to inject the database into components and get rid of the deprecated Factory::getDbo(); calls. As workflow offers already the possibility to inject the database, this pr changes the usages where a workflow is instantiated to inject the dependencies. Additionally it deprecates the usage of null dependencies and removes some dead code.

### Testing Instructions
- Enable workflow in the articles manager
- Open the article form

### Actual result BEFORE applying this Pull Request
All works.

### Expected result AFTER applying this Pull Request
All works.

### CODE REVIEW

File: libraries/src/MVC/Model/WorkflowBehaviorTrait.php [hunk 1]
```diff
@@ -77,7 +78,17 @@ public function setUpWorkflow($extension)
...
+		{
+			$db = $this->getDatabase();
+		}
+		else
+		{
+			@trigger_error('In Joomla 5 will the getDatabase function be mandatory.', E_USER_DEPRECATED);
+			$db = Factory::getContainer()->get(DatabaseDriver::class);
+		}
+
+		$this->workflow = new Workflow($extension, Factory::getApplication(), $db);
```
Comment (roland-d) [inline libraries/src/MVC/Model/WorkflowBehaviorTrait.php]: @laoneo PhpStorm gives me the warning `Expected parameter of type '\Joomla\CMS\Application\CMSApplication|void', '\Joomla\CMS\Application\CMSApplicationInterface' provided ` Is it something that could potentially cause a problem?

**Review by roland-d — CHANGES_REQUESTED** [review]

File: libraries/src/MVC/Model/WorkflowBehaviorTrait.php (hunk 1 above)
Comment (laoneo) [reply libraries/src/MVC/Model/WorkflowBehaviorTrait.php]: Not really, we try to do as much as possible with the interface. So this should be fine as the CMS app is implementing the interface.

**Review by roland-d — APPROVED** [review]

**Comment by laoneo** [comment]
Doc article is here https://github.com/joomla/Manual/pull/79

**Comment by obuisard** [comment]
Thank you Allon @laoneo !

**Pull request merged**